osutil/vfs/lists/lists.go

```diff
@@ -0,0 +1,181 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package lists implements a type-safe linked list where list nodes are
+// embedded in larger types. A single type may contain any fixed number of list
+// nodes, and thus participate in identical number of lists.
+package lists
+
+import (
+	"iter"
+	"unsafe"
+)
+
+// Node is a pair of pointers to nodes of the same type.
+//
+// Node is embedded within another type T, so that a list of values of T may be
+// created, without the necessity of separate allocation of Node structures.
+//
+// The zero value of a node is unlinked.
+type Node[T any] struct {
+	prev, next *Node[T]
+}
+
+// Unlink removes the node from the list it may be a member of.
+func (n *Node[T]) Unlink() {
+	if n.next != nil && n.prev != nil {
+		next := n.next
+		prev := n.prev
+		prev.next = next
+		next.prev = prev
+		n.next = nil
+		n.prev = nil
+	}
+}
+
+// Unlinked returns true if a node is not a member of any list.
+//
+// The zero value of a node is unlinked.
+func (n *Node[T]) Unlinked() bool {
+	return n.next == nil && n.prev == nil
+}
+
+// linkAfter arranges pointers so that [other] is after [n].
+func (n *Node[T]) linkAfter(other *Node[T]) {
+	next := n.next
```

Comment:
In kernel it's natural to use `INIT_LIST_HEAD()` but here I'm not sure what's the equivalent, so we have a case where a zero value of n is unusable for linkAfter() and linkBefore(). Since the exposed API does not allow appending/prepending for random nodes, perhaps move those to the `List` type where call to `lazyInit()` is done before modifying any structures?

Reply:
which list? `struct list_head` from https://elixir.bootlin.com/linux/v6.16.4/source/include/linux/list.h#L28?

Reply:
Yes, although the property I was referring to is on the Go side. The zero value is safe.